dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java

```diff
@@ -0,0 +1,351 @@
+import lombok.SneakyThrows;
+import lombok.val;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+@RequiredArgsConstructor
+public class GeneConstructor {
+
+  /**
+   * Constants
+   */
+  private final static Pattern QUOTES = Pattern.compile("\"");
+  private final static Pattern UNKNOWN_WHITESPACE = Pattern.compile("\\s+");
+
+  /**
+   * Dependencies
+   */
+  private final BufferedReader bufferedReader;
+  private final Map<String, String> summaryMap;
+  private final Ensembl ensembl;
+  private final GeneWriter writer;
+
+  /**
+   * State
+   */
+  ObjectNode geneNode = null;
+  ObjectNode curTranscript = null;
+  ArrayNode transcripts = DEFAULT.createArrayNode();
+  ArrayNode exons = DEFAULT.createArrayNode();
+
+  /**
+   * Method responsible for constructing the skeleton of the gene model. It streams the gtf file, constructing the
+   * elements of the gene model, and joining in information from the maps provided to the constructor of the class.
+   * Write Gene to mongo as soon as it is constructed and moves onto next one.
+   */
+  @SneakyThrows
+  public void consumeGenes() {
+    log.info("CONSUMING GENES");
+    bufferedReader.lines().filter(this::isData).map(this::parseLine).forEach(this::handleEntry);
+  }
+
+  private void handleEntry(ObjectNode entry) {
+    if (("gene".equals(asText(entry, "type")))) {
+      if (geneNode != null) {
+        finalizeAndWriteGene();
+      }
+      geneNode = constructGeneNode(entry);
+    } else if ("transcript".equals(asText(entry, "type"))) {
+      if (curTranscript != null) {
+        finalizeTranscript();
+      }
+      curTranscript = constructTranscriptNode(entry);
+    } else if ("exon".equals(asText(entry, "type"))) {
+      exons.add(constructExonNode(entry));
+    } else if ("CDS".equals(asText(entry, "type"))) {
+      ((ObjectNode) exons.get(exons.size() - 1)).put("cds", entry);
+    } else if ("start_codon".equals(asText(entry, "type"))) {
+      curTranscript.put("start_exon", exons.size() - 1);
+    } else if ("stop_codon".equals(asText(entry, "type"))) {
+      curTranscript.put("end_exon", exons.size() - 1);
+    }
+  }
+
+  private boolean isData(String line) {
+    return line.length() > 0 && line.charAt(0) != '#';
+  }
+
+  private void finalizeAndWriteGene() {
+    // Finish with the current transcript.
+    finalizeTranscript();
+    geneNode.put("transcripts", transcripts);
+    geneNode.put("external_db_ids", externalIds().get(asText(geneNode, "_gene_id")));
+    writeGene(geneNode);
+    // Reset the current state
+    curTranscript = null;
+    transcripts = DEFAULT.createArrayNode();
+  }
+
+  private void finalizeTranscript() {
+    curTranscript.put("exons", exons);
+    val trans =
+        postProcessTranscript(curTranscript, asText(geneNode, "strand"), asText(geneNode, "canonical_transcript_id"));
+    transcripts.add(trans);
+    exons = DEFAULT.createArrayNode();
+  }
+
+  private void writeGene(@NonNull ObjectNode value) {
+    writer.writeFiles(value);
```

Comment:
You're right. We should rename that method to `write` in `AbstractJongoWriter`. Thanks